crates/flox-rust-sdk/src/models/environment.rs

```diff
@@ -33,75 +35,112 @@ pub enum InstalledPackage {
     StorePath(StorePath),
 }
 
-/// Struct representing a local environment in a given location and state
+#[async_trait(?Send)]
+pub trait Environment {
+    type ConcreteTemporary;
+
+    /// Build the environment and create a result link as gc-root
+    async fn build(
+        &self,
+        nix: &NixCommandLine,
+        system: impl AsRef<str>,
+    ) -> Result<(), EnvironmentError2>;
+
+    /// Install packages to the environment atomically
+    async fn install(
+        &mut self,
+        packages: impl IntoIterator<Item = FloxPackage>,
+        nix: &NixCommandLine,
+        system: impl AsRef<str> + Send,
+    ) -> Result<&mut Self, EnvironmentError2>;
+
+    /// Uninstall remove pacakges from the environment atomically
+    async fn uninstall(
+        &mut self,
+        packages: impl IntoIterator<Item = FloxPackage>,
+        nix: &NixCommandLine,
+        system: impl AsRef<str> + Send,
+    ) -> Result<&mut Self, EnvironmentError2>;
+
+    /// Return the [EnvironmentRef] for the environment for identification
+    fn environment_ref(&self) -> &EnvironmentRef;
+
+    /// Read the catalog for this environment
+    ///
+    /// The catalog contains information about the locked sources of installed packages.
+    async fn catalog(
+        &self,
+        nix: &NixCommandLine,
+        system: impl AsRef<str>,
+    ) -> Result<EnvCatalog, EnvironmentError2>;
+
+    /// List the installed packages
+    async fn packages(
+        &self,
+        nix: &NixCommandLine,
+        system: impl AsRef<str>,
+    ) -> Result<Vec<FloxPackage>, EnvironmentError2>;
+
+    /// Create a temporary environment that can be modified freely
+    async fn modify_in(
+        &mut self,
+        path: impl AsRef<Path>,
+    ) -> Result<Self::ConcreteTemporary, EnvironmentError2>;
+
+    /// Apply the changes made in a temporary environment
+    fn replace_with(
+        &mut self,
+        temporary_environment: Self::ConcreteTemporary,
+    ) -> Result<(), EnvironmentError2>;
+
+    /// Delete the Environment
+    fn delete(self) -> Result<(), EnvironmentError2>;
+}
+
+#[async_trait]
+pub trait TemporaryEnvironment {
```

Comment:
```suggestion
pub trait TemporaryEnvironment: Environment {
```

We decided to make `TemporaryEnvironment` a super-trait of `Environment` so you can include all of the "common" functionality in `Environment`.  There's a comment on the `impl Environment for PathEnvironment<Original>` block below that explains the reasoning in a little bit more detail, I saw that first.